File: app/python/ml_tdd/src/ml/boston_predictor.py
"""Boston 住宅価格予測器モジュール."""

import os
import pickle
from typing import Optional

import numpy as np
import pandas as pd
from sklearn.linear_model import LinearRegression
from sklearn.preprocessing import StandardScaler


class BostonPredictor:
    """ボストン市の住宅価格を予測する回帰モデル.

    Attributes:
        model: 訓練済みの線形回帰モデル（未訓練時は None）
        scaler_X: 特徴量の標準化スケーラー（未訓練時は None）
        scaler_y: 目的変数の標準化スケーラー（未訓練時は None）
        train_mean: 訓練データの平均値（欠損値補完用）

    Example:
        >>> predictor = BostonPredictor()
        >>> df = predictor.load_data('data/Boston.csv')
        >>> # データの前処理と訓練（詳細は train_boston.py 参照）
    """

    def __init__(self) -> None:
        """初期化."""
        self.model: Optional[LinearRegression] = None
        self.scaler_X: Optional[StandardScaler] = None
        self.scaler_y: Optional[StandardScaler] = None
        self.train_mean: Optional[pd.Series] = None

    def load_data(self, file_path: str) -> pd.DataFrame:
        """CSV ファイルからデータを読み込む.

        Args:
            file_path: CSV ファイルのパス

        Returns:
            読み込んだ DataFrame

        Raises:
            FileNotFoundError: ファイルが存在しない場合
            ValueError: 必要な列が不足している場合
        """
        if not os.path.exists(file_path):
            raise FileNotFoundError(f"File not found: {file_path}")

        # データの読み込み
        df = pd.read_csv(file_path)

        # 必要な列の存在確認
        required_columns = ["RM", "LSTAT", "PTRATIO", "CRIME", "PRICE"]
        missing_columns = set(required_columns) - set(df.columns)
        if missing_columns:
            raise ValueError(f"Missing columns: {missing_columns}")

        return df

    def _encode_crime(self, df: pd.DataFrame) -> pd.DataFrame:
        """CRIME 列をダミー変数に変換.

        Args:
            df: 元の DataFrame

        Returns:
            CRIME がダミー変数化された DataFrame

        Note:
            drop_first=True により、最初のカテゴリ（アルファベット順）を
            基準とし、それ以外のカテゴリのダミー変数を作成
        """
        df_copy = df.copy()

        # CRIME をダミー変数化（drop_first=True）
        crime_dummies = pd.get_dummies(df_copy["CRIME"], drop_first=True, dtype=int)
        df_encoded = pd.concat([df_copy.drop("CRIME", axis=1), crime_dummies], axis=1)

        return df_encoded

    def _fill_missing_values(self, df: pd.DataFrame, fit: bool = True) -> pd.DataFrame:
        """欠損値を平均値で補完.

        Args:
            df: 対象の DataFrame
            fit: True の場合は平均値を計算して保存、False の場合は保存済みの平均値を使用

        Returns:
            欠損値が補完された DataFrame

        Raises:
            ValueError: fit=False の場合で train_mean が未設定の場合
        """
        if fit:
            # 訓練データの平均値を計算して保存
            self.train_mean = df.mean()
            return df.fillna(self.train_mean)
        else:
            # テストデータは訓練データの平均値で補完
            if self.train_mean is None:
                raise ValueError("train_mean not set. Call with fit=True first.")
            return df.fillna(self.train_mean)

    def _remove_outliers(self, df: pd.DataFrame) -> pd.DataFrame:
        """外れ値を除外.

        Args:
            df: 対象の DataFrame

        Returns:
            外れ値を除外した DataFrame

        Note:
            インデックス 76 のデータポイントを外れ値として除外
        """
        # インデックス 76 が存在する場合のみ除外
        if 76 in df.index:
            return df.drop([76], axis=0)
        return df

    def feature_engineering(self, X: pd.DataFrame) -> pd.DataFrame:
        """特徴量エンジニアリング（2乗項と交互作用項の追加）.

        Args:
            X: 元の特徴量 DataFrame

        Returns:
            エンジニアリング後の特徴量 DataFrame

        Note:
            - 2乗項: RM2, LSTAT2, PTRATIO2
            - 交互作用項: RM * LSTAT
            合計7個の特徴量を生成
        """
        X_new = X.copy()

        # 2乗項の追加
        X_new["RM2"] = X["RM"] ** 2
        X_new["LSTAT2"] = X["LSTAT"] ** 2
        X_new["PTRATIO2"] = X["PTRATIO"] ** 2

        # 交互作用項の追加
        X_new["RM * LSTAT"] = X["RM"] * X["LSTAT"]

        return X_new

    def standardize_features(self, X: pd.DataFrame, fit: bool = True) -> np.ndarray:
        """特徴量を標準化.

        Args:
            X: 特徴量 DataFrame
            fit: True の場合は fit_transform、False の場合は transform のみ

        Returns:
            標準化された特徴量（numpy 配列）

        Raises:
            ValueError: fit=False の場合で scaler_X が未設定の場合
        """
        if fit:
            self.scaler_X = StandardScaler()
            result_fit: np.ndarray = self.scaler_X.fit_transform(X)
            return result_fit
        else:
            if self.scaler_X is None:
                raise ValueError("Scaler not fitted yet. Call with fit=True first.")
            result_transform: np.ndarray = self.scaler_X.transform(X)
            return result_transform

    def standardize_target(
        self, y: pd.DataFrame, fit: bool = True
    ) -> np.ndarray:
        """目的変数を標準化.

        Args:
            y: 目的変数 DataFrame
            fit: True の場合は fit_transform、False の場合は transform のみ

        Returns:
            標準化された目的変数（numpy 配列）

        Raises:
            ValueError: fit=False の場合で scaler_y が未設定の場合
        """
        if fit:
            self.scaler_y = StandardScaler()
            result_fit: np.ndarray = self.scaler_y.fit_transform(y)
            return result_fit
        else:
            if self.scaler_y is None:
                raise ValueError("Scaler not fitted yet. Call with fit=True first.")
            result_transform: np.ndarray = self.scaler_y.transform(y)
            return result_transform

    def inverse_transform_prediction(self, y_pred: np.ndarray) -> np.ndarray:
        """予測結果を元のスケールに戻す.

        Args:
            y_pred: 標準化された予測結果

        Returns:
            元のスケールに戻された予測結果

        Raises:
            RuntimeError: scaler_y が未設定の場合
        """
        if self.scaler_y is None:
            raise RuntimeError("scaler_y not set. Train the model first.")

        result: np.ndarray = self.scaler_y.inverse_transform(y_pred)
        return result

    def train(self, X_train: np.ndarray, y_train: np.ndarray) -> None:
        """モデルを訓練する.

        Args:
            X_train: 訓練用特徴量（標準化済み）
            y_train: 訓練用目的変数（標準化済み）
        """
        self.model = LinearRegression()
        self.model.fit(X_train, y_train.ravel())

    def predict(self, X_test: np.ndarray) -> np.ndarray:
        """予測を実行する.

        Args:
            X_test: テスト用特徴量（標準化済み）

        Returns:
            予測結果（標準化済み）

        Raises:
            RuntimeError: モデルが訓練されていない場合
        """
        if self.model is None:
            raise RuntimeError("Model has not been trained yet. Call train() first.")

        result: np.ndarray = self.model.predict(X_test)
        return result

    def evaluate(self, X_test: np.ndarray, y_test: np.ndarray) -> float:
        """モデルを評価する.

        Args:
            X_test: テスト用特徴量（標準化済み）
            y_test: テスト用目的変数（標準化済み）

        Returns:
            決定係数（R²）

        Raises:
            RuntimeError: モデルが訓練されていない場合
        """
        if self.model is None:
            raise RuntimeError("Model has not been trained yet. Call train() first.")

        score: float = self.model.score(X_test, y_test)
        return score

    def save_models(
        self, model_path: str, scaler_X_path: str, scaler_y_path: str
    ) -> None:
        """モデルとスケーラーを保存.

        Args:
            model_path: モデルの保存先パス
            scaler_X_path: 特徴量スケーラーの保存先パス
            scaler_y_path: 目的変数スケーラーの保存先パス

        Raises:
            RuntimeError: モデルまたはスケーラーが未訓練の場合
        """
        if self.model is None:
            raise RuntimeError("Model has not been trained yet.")
        if self.scaler_X is None or self.scaler_y is None:
            raise RuntimeError("Scalers have not been fitted yet.")

        with open(model_path, "wb") as f:
            pickle.dump(self.model, f)
        with open(scaler_X_path, "wb") as f:
            pickle.dump(self.scaler_X, f)
        with open(scaler_y_path, "wb") as f:
            pickle.dump(self.scaler_y, f)

    def load_models(
        self, model_path: str, scaler_X_path: str, scaler_y_path: str
    ) -> None:
        """モデルとスケーラーを読み込み.

        Args:
            model_path: モデルファイルのパス
            scaler_X_path: 特徴量スケーラーファイルのパス
            scaler_y_path: 目的変数スケーラーファイルのパス

        Raises:
            FileNotFoundError: ファイルが存在しない場合
        """
        for path in [model_path, scaler_X_path, scaler_y_path]:
            if not os.path.exists(path):
                raise FileNotFoundError(f"File not found: {path}")

        with open(model_path, "rb") as f:
            self.model = pickle.load(f)
        with open(scaler_X_path, "rb") as f:
            self.scaler_X = pickle.load(f)
        with open(scaler_y_path, "rb") as f:
            self.scaler_y = pickle.load(f)

File: app/typescript/src/models/BostonPredictor.ts
import { DataFrame, IDataFrame } from 'data-forge';
import * as fs from 'fs';

interface StandardScaler {
  mean: number[];
  std: number[];
}

interface MultipleLinearRegression {
  weights: number[];
  bias: number;
}

/**
 * Boston 住宅価格予測モデル
 * 高度な回帰問題の実装（特徴量エンジニアリング + データ標準化）
 */
export class BostonPredictor {
  private model: MultipleLinearRegression | null = null;
  private scalerX: StandardScaler | null = null;
  private scalerY: StandardScaler | null = null;
  private trainMean: { [key: string]: number } | null = null;

  isTrained(): boolean {
    return this.model !== null;
  }

  /**
   * CSVファイルからデータを読み込む
   */
  async loadData(filePath: string): Promise<IDataFrame> {
    // ファイルの存在確認
    try {
      await fs.promises.access(filePath);
    } catch {
      throw new Error(`File not found: ${filePath}`);
    }

    // データの読み込み
    const fileContent = await fs.promises.readFile(filePath, 'utf-8');

    // CSV を手動でパース（CRLF と LF の両方に対応）
    const lines = fileContent.trim().split(/\r?\n/);
    const headers = lines[0].split(',').map((h) => h.trim());

    const rows = lines.slice(1).map((line) => {
      const values = line.split(',').map((v) => v.trim());
      const obj: any = {};
      headers.forEach((header, index) => {
        const value = values[index];
        // 空文字列を null に変換
        obj[header] = value === '' ? null : value;
      });
      return obj;
    });

    const df = new DataFrame(rows);

    // 必要な列の存在確認
    const requiredColumns = ['RM', 'LSTAT', 'PTRATIO', 'CRIME', 'PRICE'];
    const actualColumns = df.getColumnNames();
    const missingColumns = requiredColumns.filter(
      (col) => !actualColumns.includes(col)
    );

    if (missingColumns.length > 0) {
      throw new Error(`Missing columns: ${missingColumns.join(', ')}`);
    }

    return df;
  }

  /**
   * CRIME 列をダミー変数に変換
   * drop_first=true により、'low' を基準とし、
   * それ以外のカテゴリのダミー変数を作成
   */
  // @ts-expect-error - Used in tests via (predictor as any).encodeCrime()
  private encodeCrime(df: IDataFrame): IDataFrame {
    const uniqueCategories = df.getSeries('CRIME').distinct().toArray().sort();

    // 'low' を除外したカテゴリでダミー変数を作成
    const dummyCategories = uniqueCategories.filter((cat) => cat !== 'low');

    const encoded = df.select((row) => {
      const newRow: any = {};
      Object.keys(row).forEach((key) => {
        if (key !== 'CRIME') {
          newRow[key] = row[key];
        }
      });

      // 各ダミーカテゴリの列を作成
      dummyCategories.forEach((category) => {
        newRow[category] = row.CRIME === category ? 1 : 0;
      });

      return newRow;
    });

    return new DataFrame(encoded.toArray());
  }

  /**
   * 欠損値を平均値で補完
   */
  // @ts-expect-error - Used in tests via (predictor as any).fillMissingValues()
  private fillMissingValues(df: IDataFrame, fit: boolean = true): IDataFrame {
    if (fit) {
      // 訓練データの平均値を計算して保存
      const means: { [key: string]: number } = {};
      df.getColumnNames().forEach((col) => {
        const series = df.getSeries(col);
        const values = series
          .toArray()
          .filter((v) => v !== null && !isNaN(v as number));
        means[col] =
          (values.reduce((a, b) => (a as number) + (b as number), 0) as number) /
          values.length;
      });
      this.trainMean = means;
    }

    if (!this.trainMean) {
      throw new Error('train_mean not set. Call with fit=true first.');
    }

    const filled = df.select((row) => {
      const newRow: any = {};
      Object.keys(row).forEach((key) => {
        if (
          row[key] === null ||
          row[key] === undefined ||
          row[key] === '' ||
          isNaN(row[key] as number)
        ) {
          newRow[key] = this.trainMean![key];
        } else {
          newRow[key] = Number(row[key]);
        }
      });
      return newRow;
    });

    return new DataFrame(filled.toArray());
  }

  /**
   * 外れ値を除外
   * インデックス 76 のデータポイントを外れ値として除外
   */
  // @ts-expect-error - Used in tests via (predictor as any).removeOutliers()
  private removeOutliers(df: IDataFrame): IDataFrame {
    const dataArray = df.toArray();

    // 配列のインデックス 76 のデータポイントを除外
    const filtered = dataArray.filter((_, index) => index !== 76);

    return new DataFrame(filtered);
  }

  /**
   * 特徴量エンジニアリング（2乗項と交互作用項の追加）
   * - 2乗項: RM2, LSTAT2, PTRATIO2
   * - 交互作用項: RM_LSTAT
   * 合計7個の特徴量を生成
   */
  featureEngineering(X: IDataFrame): IDataFrame {
    const engineered = X.select((row) => {
      return {
        ...row,
        RM2: row.RM ** 2,
        LSTAT2: row.LSTAT ** 2,
        PTRATIO2: row.PTRATIO ** 2,
        RM_LSTAT: row.RM * row.LSTAT,
      };
    });

    return new DataFrame(engineered.toArray());
  }

  /**
   * 特徴量を標準化
   */
  standardizeFeatures(X: number[][], fit: boolean = true): number[][] {
    if (fit) {
      // 各列の平均と標準偏差を計算
      const numFeatures = X[0].length;
      const means: number[] = [];
      const stds: number[] = [];

      for (let i = 0; i < numFeatures; i++) {
        const col = X.map((row) => row[i]);
        const mean = col.reduce((a, b) => a + b, 0) / col.length;
        const variance = col.reduce((a, b) => a + (b - mean) ** 2, 0) / col.length;
        const std = Math.sqrt(variance);

        means.push(mean);
        stds.push(std);
      }

      this.scalerX = { mean: means, std: stds };
    }

    if (!this.scalerX) {
      throw new Error('Scaler not fitted yet. Call with fit=true first.');
    }

    // 標準化を適用
    return X.map((row) =>
      row.map((val, i) => (val - this.scalerX!.mean[i]) / this.scalerX!.std[i])
    );
  }

  /**
   * 目的変数を標準化
   */
  standardizeTarget(y: number[], fit: boolean = true): number[] {
    if (fit) {
      const mean = y.reduce((a, b) => a + b, 0) / y.length;
      const variance = y.reduce((a, b) => a + (b - mean) ** 2, 0) / y.length;
      const std = Math.sqrt(variance);

      this.scalerY = { mean: [mean], std: [std] };
    }

    if (!this.scalerY) {
      throw new Error('Scaler not fitted yet. Call with fit=true first.');
    }

    return y.map((val) => (val - this.scalerY!.mean[0]) / this.scalerY!.std[0]);
  }

  /**
   * 予測結果を元のスケールに戻す
   */
  inverseTransformPrediction(yPred: number[]): number[] {
    if (!this.scalerY) {
      throw new Error('scaler_y not set. Train the model first.');
    }

    return yPred.map((val) => val * this.scalerY!.std[0] + this.scalerY!.mean[0]);
  }

  /**
   * モデルを訓練する（多変量線形回帰）
   * 正規方程式を使用: w = (X^T X)^(-1) X^T y
   */
  train(XTrain: number[][], yTrain: number[]): void {
    // バイアス項を追加（切片）
    const XWithBias = XTrain.map((row) => [1, ...row]);

    // 転置行列を計算
    const XT = this.transpose(XWithBias);

    // XT * X を計算
    const XTX = this.matrixMultiply(XT, XWithBias);

    // 逆行列を計算
    const XTXInv = this.inverseMatrix(XTX);

    // XT * y を計算
    const XTy = this.matrixVectorMultiply(XT, yTrain);

    // w = (XT X)^(-1) XT y
    const weights = this.matrixVectorMultiply(XTXInv, XTy);

    this.model = {
      bias: weights[0],
      weights: weights.slice(1),
    };
  }

  /**
   * 予測を実行する
   */
  predict(XTest: number[][]): number[] {
    if (!this.model) {
      throw new Error('Model has not been trained yet. Call train() first.');
    }

    return XTest.map((row) => {
      const prediction =
        this.model!.bias +
        row.reduce((sum, val, i) => sum + val * this.model!.weights[i], 0);
      return prediction;
    });
  }

  /**
   * モデルを評価する（決定係数 R² を返す）
   */
  evaluate(XTest: number[][], yTest: number[]): number {
    if (!this.model) {
      throw new Error('Model has not been trained yet. Call train() first.');
    }

    const predictions = this.predict(XTest);

    // 平均値
    const yMean = yTest.reduce((a, b) => a + b, 0) / yTest.length;

    // 総平方和 (SST)
    const sst = yTest.reduce((sum, y) => sum + (y - yMean) ** 2, 0);

    // 残差平方和 (SSR)
    const ssr = yTest.reduce((sum, y, i) => sum + (y - predictions[i]) ** 2, 0);

    // 決定係数 R²
    const r2 = 1 - ssr / sst;

    return r2;
  }

  /**
   * モデルとスケーラーを保存
   */
  async saveModels(
    modelPath: string,
    scalerXPath: string,
    scalerYPath: string
  ): Promise<void> {
    if (!this.model || !this.scalerX || !this.scalerY) {
      throw new Error('Model or scalers have not been trained yet.');
    }

    const modelData = JSON.stringify(this.model, null, 2);
    const scalerXData = JSON.stringify(this.scalerX, null, 2);
    const scalerYData = JSON.stringify(this.scalerY, null, 2);

    await fs.promises.writeFile(modelPath, modelData);
    await fs.promises.writeFile(scalerXPath, scalerXData);
    await fs.promises.writeFile(scalerYPath, scalerYData);
  }

  /**
   * モデルとスケーラーを読み込み
   */
  async loadModels(
    modelPath: string,
    scalerXPath: string,
    scalerYPath: string
  ): Promise<void> {
    for (const path of [modelPath, scalerXPath, scalerYPath]) {
      try {
        await fs.promises.access(path);
      } catch {
        throw new Error(`File not found: ${path}`);
      }
    }

    const modelData = await fs.promises.readFile(modelPath, 'utf-8');
    const scalerXData = await fs.promises.readFile(scalerXPath, 'utf-8');
    const scalerYData = await fs.promises.readFile(scalerYPath, 'utf-8');

    this.model = JSON.parse(modelData);
    this.scalerX = JSON.parse(scalerXData);
    this.scalerY = JSON.parse(scalerYData);
  }

  // 行列演算のヘルパーメソッド

  private transpose(matrix: number[][]): number[][] {
    return matrix[0].map((_, colIndex) => matrix.map((row) => row[colIndex]));
  }

  private matrixMultiply(A: number[][], B: number[][]): number[][] {
    const result: number[][] = [];
    for (let i = 0; i < A.length; i++) {
      result[i] = [];
      for (let j = 0; j < B[0].length; j++) {
        let sum = 0;
        for (let k = 0; k < A[0].length; k++) {
          sum += A[i][k] * B[k][j];
        }
        result[i][j] = sum;
      }
    }
    return result;
  }

  private matrixVectorMultiply(A: number[][], v: number[]): number[] {
    return A.map((row) => row.reduce((sum, val, i) => sum + val * v[i], 0));
  }

  private inverseMatrix(matrix: number[][]): number[][] {
    // ガウスの消去法による逆行列計算（部分ピボット選択付き）
    const n = matrix.length;
    const augmented: number[][] = matrix.map((row, i) => [
      ...row,
      ...Array(n)
        .fill(0)
        .map((_, j) => (i === j ? 1 : 0)),
    ]);

    // 前進消去（部分ピボット選択）
    for (let i = 0; i < n; i++) {
      // ピボット選択：列iで最大の絶対値を持つ行を探す
      let maxRow = i;
      for (let k = i + 1; k < n; k++) {
        if (Math.abs(augmented[k][i]) > Math.abs(augmented[maxRow][i])) {
          maxRow = k;
        }
      }

      // 行を交換
      if (maxRow !== i) {
        [augmented[i], augmented[maxRow]] = [augmented[maxRow], augmented[i]];
      }

      // ピボットが0に近い場合、行列は特異行列
      if (Math.abs(augmented[i][i]) < 1e-10) {
        throw new Error('Matrix is singular or nearly singular');
      }

      // ピボット行を正規化
      const pivot = augmented[i][i];
      for (let j = 0; j < 2 * n; j++) {
        augmented[i][j] /= pivot;
      }

      // 他の行から引く
      for (let k = 0; k < n; k++) {
        if (k !== i) {
          const factor = augmented[k][i];
          for (let j = 0; j < 2 * n; j++) {
            augmented[k][j] -= factor * augmented[i][j];
          }
        }
      }
    }

    // 逆行列部分を抽出
    return augmented.map((row) => row.slice(n));
  }
}
